guide_drafts/custom_types.md

```diff
@@ -0,0 +1,155 @@
+    Ru,
+    De,
+}
+
+pub mod exports {
+    pub use super::LanguageType as Language;
+}
+```
+
+`#[postgres(type_name = "..")]` provides a name of the type as declared in our SQL, similarly `#[mysql_type = ".."]` can be used for MySql and `#[sqlite_type = ".."]` for sqlite. `FromSqlRow` derives internal types that are necessary for querying and `AsExpression` for querying. `#[sql_type = "LanguageType"]` creates a relation between a type marker and a "high-level" type for insertions as well.
+
+We are ready to implement real conversions. The first one is `ToSql`, that performs conversion to the bytes. Since it can convert our type to SQL, it allows to only make queries.
+
+```rust
+use std::io::Write;
+
+use diesel::backend::Backend;
+use diesel::serialize::{self, IsNull, Output, ToSql};
+
+impl<Db: Backend> ToSql<LanguageType, Db> for Language {
+    fn to_sql<W: Write>(&self, out: &mut Output<W, Db>) -> serialize::Result {
+        match *self {
+            Language::En => out.write_all(b"en")?,
+            Language::Ru => out.write_all(b"ru")?,
+            Language::De => out.write_all(b"de")?,
+        }
+        Ok(IsNull::No)
+    }
+}
+```
+
+That's a little bit verbose, though pretty straightforward. We just write down bytes to trait provided output, that's it. Here may be anything that the underlying database can support. So if you want to write some complicated type instead of an enum, the one should consider using a concrete backend (e.g. `diesel::pg::Pg`). And here's the implementation of `FromSql`.
+
+```rust
+use diesel::deserialize::{self, FromSql};
+use diesel::pg::Pg;
+
+impl FromSql<LanguageType, Pg> for Language {
+    fn from_sql(bytes: Option<&<Pg as Backend>::RawValue>) -> deserialize::Result<Self> {
+        match not_none!(bytes) {
+            b"en" => Ok(Language::En),
+            b"ru" => Ok(Language::Ru),
+            b"de" => Ok(Language::De),
+            _ => Err("Unrecognized enum variant".into()),
+        }
+    }
+}
+```
+
+Notice here the usage of specific backend type. It is mandatory since you need somehow to handle a generic `Backend::RawValue` (that has no trait bounds), for `Pg` in particular, it is `&[u8]` (at the diesel 1.4.4, in current master that is changed to `diesel::pg::PgValue` wrapper). Code here should be also self-explanatory though still a bit verbose.
+
+That's it! Or at least almost. That would be quite unfair to ignore a crate that generates all that stuff (though only for enums) that we dig into at this post — [diesel-derive-enum](https://crates.io/crates/diesel-derive-enum). So all of the described conversions can be generated simply with:
+
+```rust
+use diesel_derive_enum::DbEnum;
+
+#[derive(Debug, PartialEq, DbEnum)]
+pub enum Language {
+    En,
+    Ru,
+    De,
+}
+```
+
+You can find complete sources used in this post [here](https://github.com/l4l/diesel-custom-types) and leave a comment on [reddit](https://www.reddit.com/r/rust/comments/gptvej/custom_types_in_diesel/).
```

Comment:
Any chance to add this code as example [here](https://github.com/diesel-rs/diesel/tree/master/examples)?

Reply:
You mean repo with [sources](https://github.com/l4l/diesel-custom-types)? Yep, sure, add a commit to this pr?

Reply:
Yes exactly this repo. Adding it as subfolder to `examples/postgresql` should be everything that is required. 
Adding it as additional commit would be great.

Reply:
Done